crates/paimon/src/file_index/file_index_format.rs

```diff
@@ -0,0 +1,514 @@
+        let head_length = self.calculate_head_length(body_info)?;
+        let mut buffer = BytesMut::with_capacity(head_length);
+
+        // Magic
+        buffer.put_u64_le(MAGIC);
+        // Version
+        buffer.put_i32_le(Version::V1 as i32);
+        // HeadLength
+        buffer.put_i32_le(head_length as i32);
+        // ColumnSize
+        buffer.put_i32_le(body_info.len() as i32);
+
+        for (column_name, index_info) in body_info {
+            // ColumnName
+            buffer.put_u16_le(column_name.len() as u16);
+            buffer.put_slice(column_name.as_bytes());
+            // IndexTypeSize
+            buffer.put_i32_le(index_info.len() as i32);
+            // ColumnInfo，offset = headLength
+            for (index_name, IndexInfo { start_pos, length }) in index_info {
+                buffer.put_u16_le(index_name.len() as u16);
+                buffer.put_slice(index_name.as_bytes());
+                let adjusted_start = if *start_pos == EMPTY_INDEX_FLAG {
+                    EMPTY_INDEX_FLAG
+                } else {
+                    *start_pos + head_length as i32
+                };
+                buffer.put_i32_le(adjusted_start);
+                buffer.put_i32_le(*length);
+            }
+        }
+
+        // Redundant length for future compatibility
+        buffer.put_i32_le(0);
+        Ok(buffer.freeze())
+    }
+
+    fn calculate_head_length(
+        &self,
+        body_info: &HashMap<String, HashMap<String, IndexInfo>>,
+    ) -> crate::Result<usize> {
+        // Magic + Version + HeadLength + ColumnNumber + RedundantLength
+        let base_length = 8 + 4 + 4 + 4 + 4;
+        let mut total_length = base_length;
+
+        for (column_name, index_info) in body_info {
+            // Column name length + actual column name length
+            total_length += 2 + column_name.len();
+            // IndexTypeSize (index number)
+            total_length += 4;
+
+            for index_name in index_info.keys() {
+                // Index name length + actual index name length
+                total_length += 2 + index_name.len();
+                // start_pos (4 bytes) + length (4 bytes)
+                total_length += 8;
+            }
+        }
+
+        Ok(total_length)
+    }
+
+    pub fn from_file_io(file_io: FileIO, path: &str) -> crate::Result<Self> {
+        let output_file = file_io.new_output(path)?;
+        Ok(Self::new(output_file))
+    }
+}
+
+#[derive(Debug)]
+struct IndexInfo {
+    start_pos: i32,
+    length: i32,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum Version {
+    V1,
+}
+
+impl From<i32> for Version {
+    fn from(version: i32) -> Self {
+        match version {
+            1 => Version::V1,
+            _ => panic!("Unsupported file index version: {}", version),
+        }
+    }
+}
+
+impl From<Version> for i32 {
+    fn from(version: Version) -> Self {
+        match version {
+            Version::V1 => 1,
+        }
+    }
+}
```

Comment:
Maybe we can just use `#[repr(u8)]` for enum to avoid code like this?

Reply:
OK